Initial implementation/release

jtuple opened the pull request:
This pull-request is the initial implementation for `riak_ensemble`.

`riak_ensemble` is a consensus library that supports creating multiple consensus groups (ensembles). Each ensemble is a separate multi-paxos instance with it's own leader, set of members, and state.

Each ensemble also supports an extended API that provides consistent key/value operations. Conceptually, this is identical to treating each key as a separate paxos entity. However, this isn't accomplished by having each key maintain it's own paxos group -- that would be crazy. Instead, an ensemble emulates per-key consensus through a combination of per-key and per-ensemble state.

As mentioned, `riak_ensemble` supports multiple independent consensus groups. Ensembles are created dynamically, allowing applications to use `riak_ensemble` in whatever way is best fit for that application. Each ensemble also supports dynamic ensemble membership, using joint consensus to guarantee consistent membership transitions.

A given ensemble is configured to use a particular "ensemble backend".  An ensemble backend is a implemented behavior that defines how a given peer actually stores/retrieves data. For example, `riak_ensemble` ships with the `riak_ensemble_basic_backend` which stores data as an `orddict` that is saved to disk using `term_to_binary`. Where as `riak_kv` implements the `riak_kv_ensemble_backend` that is used to interface with Riak's vnodes.

Here's a link to my RICON West 2013 talk that discuss the design of `riak_ensemble` in the context of adding consistency to Riak 2.0: http://www.youtube.com/watch?v=gXJxbhca5Xg

Sibling pull-requests: basho/riak_kv#710, basho/riak_core#441

jtuple committed:
Work in progress

jtuple committed:
Remove unnecessary writes to disk

jtuple committed:
Fix peer/supervisor shutdown bug, quorum_timeout bug

jtuple committed:
Numerous changes/fixes

Fix bug in riak_ensemble_peer where quorum_timeout messages were
being ignored, rather than being passed on to riak_ensemble_msg
for translation into {timeout, Replies} messages that the peer
was expecting.

Add sync_request/sync behavior to ensemble backends and implement
for riak_ensemble_basic_backend. Adding syncing has been a long
standing feature addition, but turns out is necessary for the
root ensemble case in order to guarantee empty root peers learn
root state before potentially taking over. Newly started peers
are untrusted, and only become trusted after syncing with a majority
of peers across all current views.

Change how riak_ensemble_peer to delay transitioning views until
after the larger joint membership is following the leader, this
ensures that at least a majority of newly added peers became trusted
before the transition.

Fix bug related to handling view_seq for check_ensemble. The view_seq
is now set to the fact epoch/seq at which the new view was created,
as well as incremented when a new epoch is established. This ties
view consistency to fact consistency.

Add delay between a probe timeout and retrying to probe.

jtuple committed:
Remove hardcoded old_hash define

jtuple committed:
Minor clean-up

jtuple committed:
Remove debug messages before tech preview

jtuple committed:
Timeout requests quickly when target node offline

jtuple committed:
Update sc.erl test for bucket types

andrewjstone commented on Makefile:
```diff
@@ -0,0 +1,36 @@
+	@echo
+	dialyzer --output_plt $(DEPS_PLT) --build_plt --apps $(DEPS) -r deps
+
+dialyzer: $(DEPS_PLT)
```
For this to work there needs to be a deps directory, however `./rebar get-deps` won't create it since there aren't any deps.

jtuple replied:
Added empty `deps/.gitignore` to ensure we always have a deps directory present.

andrewjstone commented on Makefile:
```diff
@@ -0,0 +1,36 @@
+eunit: compile
+	$(REBAR) skip_deps=true eunit
+
+test: compile eunit
```
This should be removed

jtuple replied:
Keeping this around, since we'll inevitably add more eunit tests. Also, want to update `test/sc.erl` to actually work properly as an eunit test. In any case, fixed `test/sc.erl` to build cleanly (so that `make test` no longer fails), as well as changed the EQC test to currently print "Test disabled" and fake pass. The test is mostly designed to be driven by a `riak_test` (work still pending), but I want it to also work as an eunit as well. WIll be future PR to fix that. But, for now, `make test` compiles/passes.

andrewjstone commented on src/riak_ensemble_basic_backend.erl:
```diff
@@ -0,0 +1,188 @@
+                        {ok, binary_to_term(Binary)}
+                    catch
+                        _:_ ->
+                            not_found
```
Do we really want to treat a corrupted file the same as one that doesn't exist? We may at least want to log an error.

jtuple replied:
Added warning message. Which, sadly, also meant bringing in lager as dependency (wanted to keep `riak_ensemble` generic and not depend on lager, but I'm too lazy to deal with things now and took the easy route). We do want to start with an empty backend state though, since we're designed to handle that case. The peer will come up untrusted, but after syncing with other peers, will have a perfectly valid state once again (assuming a majority of remaining have valid states -- which is an inherent assumption in `riak_ensemble` anyway).

andrewjstone commented on src/riak_ensemble_router.erl:
```diff
@@ -0,0 +1,252 @@
+    throw("infinity timeout not currently safe for non-pid target");
+sync_send_event(Node, Target, Event, Timeout) ->
+    Ref = make_ref(),
+    spawn_link(?MODULE, sync_proxy, [self(), Ref, Node, Target, Event, Timeout]),
```
Should this really be a spawn_link? Do we want to crash the caller (most likely a client) in this case?

jtuple replied:
As written, we _must_ crash. We're waiting infinitely for a message to be received from the spawned proxy process. If that process ever dies (due to some bug we introduce in `sync_proxy`), we would wait forever. Remember, this is (at worst), emulating the behavior of `gen_fsm:sync_send_event` which crashes the caller in various scenarios.

andrewjstone commented on src/riak_ensemble_router.erl:
```diff
@@ -0,0 +1,252 @@
+
+fail_cast({sync_send_event, From, Ref, _Event, _Timeout}) ->
+    From ! {Ref, timeout};
+fail_cast(_) ->
```
AFAICT this function only gets called from one place with the form above. This catchall clause seems like a good place to fail fast instead

jtuple replied:
Changed.

andrewjstone commented on src/riak_ensemble_router.erl:
```diff
@@ -0,0 +1,252 @@
+            ok
+    end.
+
+-spec cast(ensemble_id(), msg()) -> error | ok.
```
This function is a wrapper and only called inside this module. It should just be removed IMO.

andrewjstone commented on src/riak_ensemble_router.erl:
```diff
@@ -0,0 +1,252 @@
+
+-spec cast(node(), ensemble_id(), msg()) -> error | ok.
+cast(Node, Ensemble, Msg) when Node =:= node() ->
+    cast(Ensemble, Msg);
```
call ensemble_cast/2 directly here instead and remove cast/2

andrewjstone commented on src/riak_ensemble_router.erl:
```diff
@@ -0,0 +1,252 @@
+
+-spec handle_ensemble_cast(_,_) -> ok.
+handle_ensemble_cast({sync_send_event, From, Ref, Event, Timeout}, Pid) ->
+    spawn(fun() ->
```
Can just call sync_proxy_direct/5 from inside this spawn. The code is the same except for the return of ok. Not sure if this is here because of performance or just refactoring needed.

andrewjstone commented on src/riak_ensemble_router.erl:
```diff
@@ -0,0 +1,252 @@
+
+-spec sync_proxy_router(pid(), reference(), node(), ensemble_id(), msg(), timeout()) -> ok.
+sync_proxy_router(From, Ref, Node, Target, Event, Timeout) ->
+    case riak_ensemble_router:cast(Node, Target, {sync_send_event, self(), Ref, Event, Timeout}) of
```
Probably don't need to fully qualify this unless planning for hot code upgrade.

andrewjstone commented on src/riak_ensemble_router.erl:
```diff
@@ -0,0 +1,252 @@
+                    handle_ensemble_cast(Msg, Pid),
+                    ok;
+               true ->
+                    riak_ensemble_router:cast(Node, Ensemble, Msg),
```
Probably don't need to fully qualify this unless planning for hot code upgrade.

andrewjstone commented on src/riak_ensemble_manager.erl:
```diff
@@ -0,0 +1,543 @@
+
+-spec root_set_ensemble_once(ensemble_id(), ensemble_info()) -> std_reply().
+root_set_ensemble_once(EnsembleId, Info) ->
+    riak_ensemble_peer:kmodify(node(), root, ensembles,
```
Should probably use rmodify here for consistency

andrewjstone commented on src/riak_ensemble_manager.erl:
```diff
@@ -0,0 +1,543 @@
+    ThisNode = node(),
+    case node(Pid) of
+        ThisNode ->
+            %% TODO: Remove peer reference here, no?
```
If you remove it from ets, which seems to be the correct thing, the new one will need to get added when the peer restarts. It shouldn't be part of remote_peers.

andrewjstone commented on src/riak_ensemble_peer.erl:
```diff
@@ -0,0 +1,1416 @@
+%%%===================================================================
+
+-spec init([any(),...]) -> {ok,election | probe,state()}.
+init([Mod, Ensemble, Id, Views, Args]) ->
```
I don't think `Views` needs to be passed in. It can be reloaded from the saved fact or undefined. Maintaining peer state in both `riak_ensemble_manager` and `riak_ensemble_peer` seems like it could be dangerous.

andrewjstone commented on src/riak_ensemble_peer.erl:
```diff
@@ -0,0 +1,1416 @@
+                    C),
+    ETS = ets:new(x, [public, {read_concurrency, true}, {write_concurrency, true}]),
+    Saved = reload_fact(Ensemble, Id),
+    Fact = Saved#fact{views=Views},
```
If you don't pass in Views, you can just use Saved.

andrewjstone commented on src/riak_ensemble_manager.erl:
```diff
@@ -0,0 +1,543 @@
+                 #ensemble_info{mod=Mod, args=Args, members=Bootstrap} = Info,
+                 %% Start new peer
+                 %% io:format("Should start: ~p~n", [{Ensemble, Id, Bootstrap, Mod, Args}]),
+                 %% TODO: Make Bootstrap be views not membership
```
If you don't pass in Bootstrap, you don't have to worry about this

andrewjstone commented on src/riak_ensemble_peer.erl:
```diff
@@ -0,0 +1,1416 @@
+            ViewSeq = {Fact#fact.epoch, Fact#fact.seq},
+            NewFact = Fact#fact{views=Views2, view_seq=ViewSeq},
+            pause_workers(State),
+            case try_commit(NewFact, State) of
```
This call to try_commit will fail with a timeout when the number of new members is a quorum of the new view. This occurs because the following peers are not yet alive because they haven't been started by `riak_ensemble_manager` on their respective nodes. They haven't been started because `riak_ensemble_manager` hasn't learned of them yet. 

There are actually two issues here. The first issue is a bug. The new fact should only be committed to the original view and hence it should never actually timeout, as it will just be committing to it's existing views. Right now a commit is attempted to all views, including ones that aren't yet committed.

The second issue, is that even if we commit locally, on the next tick we will try to commit to the new peers. We will run into the same issue where they aren't alive yet. The current proposal to 'fix' this is to add a new async api to the manager that will broadcast in a best effort manner, any new peers to start to all managers. It won't change the state of the ensembles, just start peers that 'might be needed soon', by a commit of a new view for instance. If this occurs in time then the leader will maintain it's leadership. Otherwise, it will drop down to probe state and the ensemble state will propogate in the normal manner. 

When making these changes it's important to ensure that the peers cannot get into a state where a leader isn't elected and hence the state cannot propogate to start peers that require a leader to be elected. This would cause the system to become unavailable.

andrewjstone replied:
Thinking about the code some more, I'm thinking that it's NOT a bug to commit to the new members. The race for starting the peers still exists though. The reason we want to commit to the new members as well is to prevent a no leader situation if the current leader isn't a part of the new view. The peers also need that information to know who to probe if the leader goes away.

andrewjstone commented on src/riak_ensemble_basic_backend.erl:
```diff
@@ -0,0 +1,188 @@
+                       riak_ensemble_backend:latest_obj(?MODULE, Obj1, Obj2)
+               end,
+    Data2 = lists:foldl(fun({_PeerId, PeerData}, Acc) ->
+                                orddict:merge(LatestFn, Acc, PeerData)
```
If an old value has been deleted in a later epoch, won't this merge result in the value being ressurected?

And does it actually matter, since the rewrite will occur if read in a new epoch?

jtuple replied:
`riak_ensemble` currently never deletes data. We always write a tombstone value that is never reaped. We must to handle the partial write during network partition case. There are some plans to address this in the future, but the entire protocol today assumes we never delete things.

andrewjstone commented on src/riak_ensemble_peer.erl:
```diff
@@ -0,0 +1,1416 @@
+    {next_state, following, State3};
+following({prepare, Id, NextEpoch, From}=Msg, State=#state{fact=Fact}) ->
+    Epoch = epoch(State),
+    case (Id =:= leader(State)) and (NextEpoch > Epoch) of
```
Why does the follower need to ensure a prepare with a higher epoch is from the same leader?

Will this mean that a new leader can never get elected without the actual old one failing and all the current followers going back to probe state first?

jtuple replied:
Yes. We have stable leader leases. That's specific design decision in `riak_ensemble`. A new leader can only be elected after folks have given up on the old leader, thus two leaders can never be leading at once, ever. This is required to allow the leader to respond directly for reads that are from the current epoch, without needing to contact other peers.

andrewjstone commented on src/riak_ensemble_peer.erl:
```diff
@@ -0,0 +1,1416 @@
+    {next_state, leading, State2};
+leading_kv({put, Key, Fun}, From, State) ->
+    Self = self(),
+    async(Key, State, fun() -> do_put_fsm(Key, Fun, From, Self, State) end),
```
I presume these asynch operations work properly and are applied in the right (seq number) order because each key goes to the same worker. Is that correct? Otherwise, we'd probably have to verify validity of each request by seq number.

jtuple replied:
That's the assumption, yes. Do we need to be more paranoid?

andrewjstone replied:
Nope. that was just me trying to figure this out. The way it works sounds good considering the guarantees of erlang messaging and tcp in the future.

andrewjstone commented on src/riak_ensemble_peer.erl:
```diff
@@ -0,0 +1,1416 @@
+    Obj2 = increment_obj(Key, Obj, State),
+    Peers = get_peers(Members, State),
+    {Future, State2} = blocking_send_all({put, Key, Obj2, Id, Epoch}, Peers, State),
+    %% TODO: local can be failed here, what to do?
```
It is probably safest to crash here. If local is unavailable we may not want to trust the data when it comes back. A restart would make it sync again.

Alternatively we could just step down and go directly to probe or sync state.

jtuple replied:
Yes. Created issue #2 to track this. Will fix/test in future pull-request.

jtuple committed:
Address code review comments/concerns

Update Makefile to check for rebar, as well as use less strict dialyzer
parameters by default.

Make test/sc.erl compile clean without warnings so that 'make test' works.

Fix an 'ummatched return' dialyzer complaint.

Add warning for corrupted riak_ensemble_basic_backend state.

Remove catch-all case for riak_ensemble_router:fail_cast.

jtuple commented:
Address many of the review comments. Plan to do additional code clean-up in a future pull-request, but want to merge this so that we can start basing new pull-requests against develop.

Many of the other comments here were discussed in backchannels before. Such as why we pass `View` into the peers. Although, I vaguely remember some discussion on how we can remove some redundancy.

The issue with spuriously stepping down when changing membership is also something we've discussed in length, and something I believe the [ajs-wip-2.0-fixes](https://github.com/basho/riak_ensemble/tree/ajs-wip-2.0-fixes) branch is designed to fix. We should wrap that up and merge in when we get a chance.

Any comments on what remains given this review? Any objections to merging this pull-request to give `riak_ensemble` an actual develop branch people can base pull-requests off of, and to start fixing remaining items in new pull-requests?

andrewjstone commented:
:+1:  merge it in
